homeassistant/components/roborock/rest.py

```
@@ -0,0 +1,54 @@
+"""Roborock Rest API."""
+
+import logging
+
+from roborock.containers import HomeDataDevice, HomeDataScene, UserData
+from roborock.exceptions import RoborockException
+from roborock.web_api import RoborockApiClient
+
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class RoborockRestApi:
+    """Class to manage fetching data from the rest API."""
+
+    def __init__(
+        self,
+        device: HomeDataDevice,
+        api_client: RoborockApiClient,
+        user_data: UserData,
+    ) -> None:
+        """Initialize."""
+        self._device = device
+        self._user_data = user_data
+        self._api_client = api_client
+
+    async def get_scenes(self) -> list[HomeDataScene]:
+        """Get scenes."""
+        try:
+            return await self._api_client.get_scenes(self._user_data, self._device.duid)
+        except RoborockException as err:
+            _LOGGER.warning(
```

This does not seem like appropriate error handling to just log warnings and return an empty result like nothing happened.

---

If we're doing this because not all devices support scenes then we should probably figure out which devices support them?

---

So doing a test on my side - my device does not support scenes, and it instead just gives an empty list.

So RoborockException would not be valid as a catch for not supported devices and definitely shouldn't be ignored as it is now.

So I think calling get scenes and then setting up the entities will work - as we will get a empty list for devices that don't support it and therefore setup no entities. Figuring out explicitly what devices support scenes will likely be a harder task, and I think since it just returns an empty list, I would think it would be fine just to not setup any entities.

---

Ok I will remove. The thought behind it was to not cause the device to fail to setup just because of scenes but now that the logic to handle it is contained in the entity (and not init) it doesn't really matter.

---

Yep, understood the thought and intention.  

If there was a transient error that was just swallowed and we continue anyway, then the user would just randomly miss entities which isn't desirable.

There are specific best practices for retrying errors with backoff, etc, and incidicating status about the integration or entity, and error message to the user so raising HomeAssistantError is the way to go here.   If its a non-retryable error (e.g. bad request) then that root cause would need to be fixed to not happen in the first place.